components/external_storage/src/lib.rs

```diff
@@ -77,10 +79,16 @@ pub trait ExternalStorage: 'static + Send + Sync {
         storage_name: &str,
         restore_name: std::path::PathBuf,
         expected_length: u64,
+        expected_sha256: Option<Vec<u8>>,
         speed_limiter: &Limiter,
         file_crypter: Option<FileEncryptionInfo>,
     ) -> io::Result<()> {
         let reader = self.read(storage_name);
+        if let Some(p) = restore_name.parent() {
+            // try create all parent dirs from the path (optional).
+            // ignore the error because once it failed, it would soon be revealed by succeed steps.
+            let _ = fs::create_dir_all(p);
```

Comment:
That is because the document of `create_dir_all` is (someway) unclear about the behavior when the dir already exists.
I am afraid if we fail fast here, in some implement we may fail when the dir already exists. 🤔

Reply:
If there is an error, `create_dir_all` and following fs operation fails either ways, why not fail fast? So we know which causes error.

----

> I am afraid if we fail fast here, in some implement we may fail when the dir already exists. 🤔

We can ignore allowed errors.